### PR TITLE
Bound unknown profile metadata retained in directory caches

### LINKED ISSUE #867: marmot-app: directory profile `extra` fields bypass the per-field ingest cap — unbounded attacker-controlled kind:0 metadata persisted to the directory cache
**Severity:** LOW (memory/storage bloat; defeats an explicit, documented mitigation)
**Component:** `crates/marmot-app`

## Summary

Nostr kind:0 metadata is fully attacker-controlled (anyone can publish any content to a relay). The directory-record parser deliberately caps each **known** profile string field at `MAX_PROFILE_FIELD_CHARS = 4096` and caps follows at `MAX_FOLLOW_LIST_ENTRIES`, with an explicit comment that the intent is to "keep a malicious multi-megabyte value from bloating the directory cache and downstream consumers." That mitigation is silently defeated for **unknown** fields: `extra_profile_fields` copies every non-known JSON key and value **verbatim** with no per-value size cap and no key-count cap, and the resulting `extra` map is `#[serde(flatten)]`-serialized into `profile_json` and persisted to the (SQLCipher) per-account directory cache and the shared public-directory store.

## Location

- `crates/marmot-app/src/directory/records.rs:264-273` — `extra_profile_fields`: `object.iter().filter(|(key,_)| !is_known_profile_field(key)).map(|(key,value)| (key.clone(), value.clone())).collect()` — no size/count bound.
- Contrast the cap at `:290-305` (`MAX_PROFILE_FIELD_CHARS` / `string_field`, which truncates known fields).
- Persistence: `UserProfileMetadata.extra` is `#[serde(flatten)]` (`records.rs`), serialized into `profile_json` and stored via the directory cache / shared storage.

## Failure scenario

A hostile user publishes a kind:0 whose content places a large value under an unknown key (e.g. `{"banner":"<10 MB string>"}` or a deeply nested unknown object). When any account ingests that profile — `refresh_directory_profiles`, `fetch_current_user_profile_for_account_id`, or the live directory-sync path `ingest_directory_relay_event → profile_from_record → remember_directory_profile_if_newer` — the full blob is persisted to that account's directory cache and the shared store, and re-materialized in memory on every `directory_entries()` / search / notification-user lookup. Repeated across many hostile pubkeys this bloats the encrypted cache without bound, on a path whose whole point was to be bounded.

## Suggested fix

Bound `extra` on ingest, mirroring the `MAX_PROFILE_FIELD_CHARS` / `MAX_FOLLOW_LIST_ENTRIES` posture: cap the number of retained unknown keys, and drop (or truncate) any entry whose serialized value exceeds a size limit. Simplest: in `extra_profile_fields`, skip entries whose `value` serializes beyond a cap and stop after N keys.

## Notes

Upstream relay event-size limits may bound the worst case in practice, but they are not guaranteed and the code's own stated mitigation is defeated. This is on-disk **size** bloat, not a plaintext-exposure issue (distinct from the closed #682, which made the directory cache encrypted).

---
<sub>Filed from a meticulous multi-agent code-review pass; verified against source and deduplicated against open/closed issues.</sub>

### PR DESCRIPTION
**Pull request by agent-p1p**
## Problem

Known profile fields already have retention limits, but unknown kind:0 fields were copied into directory caches without size or count bounds. Oversized remote metadata could therefore bloat stored profiles and downstream reads.

## Solution

Bound unknown fields in the shared profile-ingestion parser:

- Retain at most 32 fields, with compact-JSON limits of 256 bytes per key and 4,096 bytes per value.
- Drop oversized entries whole; measure encoded size without allocating another attacker-sized serialized copy.
- Preserve supported profile fields and normal custom-field round trips.
- Cover parser boundaries, cache persistence, search, fetch, and refresh paths with regression tests.

[Implementation plan](https://github.com/marmot-protocol/mdk/issues/867#issuecomment-5574608033)

Fixes #867

## Scope and limitations

These limits apply when profiles are ingested again; this change does not shrink existing cached rows immediately. Bounding the initial raw event JSON parse is a separate hardening task and is not included here.

<details>
<summary>Builder-reported checks and known baseline failures</summary>

- cargo fmt -p marmot-app: passed
- git diff --check: passed
- cargo clippy --locked -p marmot-app --all-targets -- -D warnings: passed
- cargo test --locked -p marmot-app --lib directory::records::tests: 14 passed
- cargo test --locked -p marmot-app --lib directory::search::tests: 26 passed
- cargo test --locked -p marmot-app --lib ingesting_kind0_profile_persists_only_bounded_unknown_fields: passed
- cargo test --locked -p marmot-app --lib unknown_kind0_fields: 2 passed
- cargo test --locked -p marmot-app --lib remember_directory_profile_if_newer_keeps_local_edit_on_equal_timestamp: passed
- cargo test --locked -p marmot-app --test relay_runtime fetch_and_refresh_profile_inherit_extra_field_bounds_from_shared_parser: passed
- cargo test --locked -p marmot-app --test relay_runtime runtime_profile_publish_preserves_unknown_kind0_fields: passed
- cargo test --locked -p marmot-app --lib: 1004 passed, 4 failed, 2 ignored; three epoch-stall tests also fail on planned base a65c1fa72a984d1893edbb6be575384ae8725008 without this change; account_reconcile_returns_local_readiness_before_relay_subscription_registration passed on serial retry
- cargo test --locked -p marmot-app --test relay_runtime: 148 passed, 5 failed, 1 ignored; sampled encrypted-media and retention failures also fail on planned base a65c1fa72a984d1893edbb6be575384ae8725008 without this change
- just fast-ci: passed

</details>

## Summary by CodeRabbit

- **Bug Fixes**
  - Limited retention of unknown profile fields to prevent oversized or excessive metadata from being stored.
  - Oversized, deeply nested, or invalid extra fields are now omitted while supported profile information remains available.
  - Applied consistent profile-field limits across ingestion, search results, caching, fetching, refreshing, and reopened sessions.
  - Preserved recognized profile fields and endpoint-derived information while excluding conflicting or untrusted duplicates.
  - Ensured refreshed profiles consistently reflect newer metadata while retaining valid extra fields.

### CODE REVIEW

**Comment by agent-p1p** [comment]
## Pip shadow disposition: ready for human review

Pip case `repo:1055628515#867@3` at state revision 12 completed its deterministic final review on `378d3a69a89b866ce767caa18172c3190fc00b2e`. Autonomous merge is disabled; a human merge decision is required.

File: crates/marmot-app/src/directory/search.rs [hunk 1]
```diff
@@ -1498,6 +1500,123 @@ mod tests {
         );
     }
 
+    /// Hostile kind:0 extensions on a relay-resolved stranger must inherit the
+    /// shared ingest bounds in both the streamed result and the un-promoted
+    /// search-graph cache. The stranger is still not promoted.
+    #[tokio::test]
+    async fn a_relay_resolved_search_profile_inherits_extra_field_bounds() {
```
Comment (erskingardner) [inline crates/marmot-app/src/directory/search.rs]: This test is a near-clone of `a_profile_resolved_from_a_relay_is_cached_without_promoting_the_stranger`, which sits directly above it at `search.rs:1426`: same `MockRelay::run`, same stranger identity plus `wait_for_network_ready` and `publish_user_profile`, same follower cache seed, same `search_users`/`drain`, same `search_record` read, and the same `cache.entry(...).is_none()` assertion carrying the same message string. The only new information is three assertions about `extra` (`website`/`bot` kept, `custom_blob` dropped).

Publishing the hostile `custom_blob` from that existing test's `UserProfileMetadata` and adding those three assertions there buys the same coverage for roughly 15 lines instead of 123, and keeps one mock relay in this module instead of two. The added `directory_sync_plan` assertion also re-covers mdk#418, which the existing test's `cache.entry` guard already owns.

File: crates/marmot-app/src/directory/records.rs [hunk 2]
```diff
@@ -772,4 +833,358 @@ mod tests {
...
+        .expect("object kind:0 content parses")
+        .1
+    }
+
+    fn ascii_xs(count: usize) -> String {
+        "x".repeat(count)
+    }
+
+    /// Braces, encoded keys, colons, values, and commas at maximum occupancy.
+    const MAX_RETAINED_EXTRA_PROFILE_JSON_BYTES: usize = 2
```
Comment (erskingardner) [inline crates/marmot-app/src/directory/records.rs]: `MAX_RETAINED_EXTRA_PROFILE_JSON_BYTES` is copy-pasted verbatim into three modules (here, `directory/search.rs:1512`, and `src/tests.rs:10347`), and every assertion against it is vacuous. The bound is `2 + 32 * (256 + 1 + 4096) + 31` = 139,296 bytes, while the maps actually asserted are about 290 bytes (`{"k00":0,…,"k31":31}` at line 1081) or two-to-three small fields in the other two tests. Since those same tests already assert per-entry acceptance and rejection, these checks cannot fail.

Either drop the constant and its three assertions, or -- if the aggregate ceiling is the property worth pinning -- declare it once beside `MAX_EXTRA_PROFILE_VALUE_BYTES` and assert it against a genuinely saturated map (32 entries with 254-byte keys and 4094-byte values). That is the only shape that would catch a regression in the derivation. As written it is about 15 duplicated lines carrying no signal.

File: crates/marmot-app/src/directory/records.rs [hunk 3]
```diff
@@ -338,15 +339,73 @@ pub(crate) fn profile_from_record(
...
+/// including string quotes/escapes and all nested containers and member keys.
+/// Oversized values are dropped whole, never truncated.
+pub(crate) const MAX_EXTRA_PROFILE_VALUE_BYTES: usize = 4096;
+
+/// Counting `Write` sink that accepts compact JSON only while it fits `remaining`.
+///
+/// The next write that would exceed the budget fails immediately without
+/// accepting a prefix or counting those bytes. Measurement never materializes a
+/// serialized attacker-sized copy.
+struct BoundedJsonWriteBudget {
```
Comment (erskingardner) [inline crates/marmot-app/src/directory/records.rs:368]: The doc comment's justification does not hold. By the time `extra_profile_fields` runs, both `record.event.content` (the raw attacker-controlled `String`) and the fully parsed `serde_json::Value` are already resident, so a `serde_json::to_vec(value)` would be a third transient copy of data already held twice -- not the difference between bounded and unbounded memory. The whole helper could be:

```rust
fn compact_json_within_budget<T: Serialize + ?Sized>(value: &T, budget: usize) -> bool {
    serde_json::to_vec(value).is_ok_and(|encoded| encoded.len() <= budget)
}
```

which removes the struct, the `Write` impl, the new `std::io` import, and `bounded_json_write_budget_rejects_the_write_that_would_exceed` -- about 55 lines.

The sink does buy a real CPU short-circuit: a 2 MB value is rejected on its first `write_all` instead of being re-encoded in full. That makes keeping it defensible. If you keep it, please reword the comment to claim the short-circuit rather than the memory property, so a later reader does not treat the avoided copy as load-bearing.

File: crates/marmot-app/src/directory/records.rs [hunk 4]
```diff
@@ -772,4 +833,358 @@ mod tests {
...
+        let encoded = serde_json::to_vec(&over_profile.extra).unwrap();
+        assert!(encoded.len() <= MAX_RETAINED_EXTRA_PROFILE_JSON_BYTES);
+        assert_eq!(
+            profile_content_json(&over_profile)["name"],
+            serde_json::json!("typed")
+        );
+    }
+
+    #[test]
+    fn extra_profile_fields_do_not_enable_unbounded_json_recursion() {
```
Comment (erskingardner) [inline crates/marmot-app/src/directory/records.rs]: This test asserts serde_json's own default `RECURSION_LIMIT` (128) rejects 200-deep nesting, and then that `profile_from_record` therefore returns `None`. Both hold identically on the base commit -- nothing in this PR touches the parse -- so it is not a regression test for the new bound.

If the intent is a tripwire against someone later reaching for `Deserializer::disable_recursion_limit()`, that is a reasonable thing to pin, but say so in the doc comment and name the test for the parse rather than for `extra_profile_fields`, which it never exercises. Otherwise it is coverage of a dependency default.

**Review by erskingardner — DISMISSED** [review]
| Review metadata | Value |
| --- | --- |
| Reviewed at (UTC) | `2026-09-07T20:40:31Z` |
| Commit reviewed | `378d3a69a89b866ce767caa18172c3190fc00b2e` |
| Model | `claude-opus-5[1m]` |
| Reasoning level | `Not exposed by runtime` |
| Recommended action | **Merge** |

## Summary

The production change is right, and it is small: 40 lines placed exactly at the root cause (`extra_profile_fields`), restoring the posture that `MAX_PROFILE_FIELD_CHARS` and `MAX_FOLLOW_LIST_ENTRIES` already document a few lines below. Dropping a whole entry rather than truncating is the correct call for arbitrary JSON; checking the key before visiting the value avoids serializing oversized values; known and rejected entries correctly do not consume quota; and `remaining` cannot underflow. `serde_json::Map` is a `BTreeMap` in this workspace (no `indexmap` in `Cargo.lock`, so `preserve_order` is off), so "retain the first 32" is deterministic. Worst-case retained extras (about 136 KB) land in the same order of magnitude as the existing known-field ceiling (7 char-capped fields, up to roughly 114 KB of 4-byte UTF-8), so the new budget is proportionate rather than a new dominant term. Pre-existing oversized rows also self-repair on the next newer kind:0, since `remember_directory_profile_if_newer` replaces the profile wholesale.

All of my concerns are in the test payload, and none of them block. 736 added lines for a 40-line fix is a lot, and the property under test is enforced in a single pure function that every caller funnels through -- `profile_from_record` is the only non-test initializer of `extra`, reached from `records.rs:595`, `methods.rs:1288`, `search.rs:638`, and `search.rs:800`.

## Non-blocking findings

- `directory/search.rs:1507` is a near-clone of the test immediately above it; the three new `extra` assertions belong in that existing test (about 15 lines instead of 123, and one mock relay in the module instead of two).
- `MAX_RETAINED_EXTRA_PROFILE_JSON_BYTES` is triplicated across three modules and every assertion on it is vacuous: a 139,296-byte bound checked against roughly 290-byte maps.
- `BoundedJsonWriteBudget`'s doc comment claims a memory benefit that does not hold once the content `String` and parsed `Value` are already resident. The CPU short-circuit is the real (and smaller) win, and worth saying instead.
- `extra_profile_fields_do_not_enable_unbounded_json_recursion` pins serde_json's default recursion limit and passes identically on the base commit.

The `relay_runtime.rs` test does earn its place for the fetch half: a foreign signed publisher is the one thing the app's own publish path cannot simulate, given the equal-timestamp local-cache masking the plan called out. Its second half -- a whole second app and tempdir for `refresh_profile_for_account_id` -- reaches the same parser via `latest_fresh_profiles_from_records` and could go for about 40 lines with little loss.

One acknowledged residual, already raised by the security reviewer and in #867 itself: `profile_from_record` still `from_str`s the full attacker-controlled content before any bound applies, so transient memory stays unbounded. That is persistence-versus-memory and correctly outside this PR's scope, but it is the natural follow-up if #867's LOW rating ever rises.

All required CI checks pass at this SHA.

File: crates/marmot-app/src/directory/search.rs (hunk 1 above)
Comment (erskingardner) [reply crates/marmot-app/src/directory/search.rs]: Addressed in d14066be: consolidated this into the existing search-cache test, retaining the streamed/cached unknown-field assertions and the non-promotion guard. Removed the duplicate relay/app setup and redundant sync-plan assertion.

File: crates/marmot-app/src/directory/records.rs (hunk 4 above)
Comment (erskingardner) [reply crates/marmot-app/src/directory/records.rs]: Addressed in d14066be: renamed the test to profile_parse_preserves_the_json_recursion_limit and documented it as an intentional parser-default tripwire. It explicitly fails before extra-field retention and is not presented as a regression test for the new retention cap.

File: crates/marmot-app/src/directory/records.rs (hunk 2 above)
Comment (erskingardner) [reply crates/marmot-app/src/directory/records.rs]: Addressed in d14066be: replaced the loose checks with one genuinely saturated 32-entry map test and added escaped/Unicode key-boundary coverage. The complete encoded map is 139,329 bytes: 139,296 entry bytes plus 33 bytes for braces and commas. Removed the duplicated constants and tiny-map ceiling assertions.

File: crates/marmot-app/src/directory/records.rs (hunk 3 above)
Comment (erskingardner) [reply crates/marmot-app/src/directory/records.rs:368]: Addressed in d14066be: kept the streaming sink to stop subsequent serialization writes and avoid an additional encoded buffer, but clarified that it does not bound the raw event or already-parsed Value. The documentation also notes that string scanning can precede the rejected write; it makes no total-memory or constant-time claim.

**Comment by erskingardner** [comment]
Addressed the review suggestions in d14066be (tests and documentation only; no production behavior changes).

- Added escaped/Unicode key edge cases and an exact 139,329-byte saturated-map test.
- Consolidated duplicate search setup; reused fetch/refresh setup while proving that refresh consumes a newer, independently signed event.
- Clarified the counting sink and parser-recursion test, and documented that historical cache rows retain their old contents until newer metadata replaces them.

The initial raw-event parse cap remains deferred: it needs an explicit event-size/rejection policy and is separate from this retained-storage fix. This PR does not introduce automatic cleanup of historical cache rows.

Validation: all 43 targeted tests and `just fast-ci` passed locally. Fresh CI is running on d14066be; the earlier reviews assessed 378d3a69.

**Review by Datawav — APPROVED** [review]

**Review by erskingardner — APPROVED** [review]
| Review metadata | Value |
| --- | --- |
| Reviewed at (UTC) | `2026-09-08T05:56:22Z` |
| Commit reviewed | `d14066be1b5355b9ba1c901ba53b30c8cdb2dec3` |
| Model | `gpt-6-astra` |
| Reasoning level | `high` |
| Recommended action | **Merge** |

No blocking findings. The shared ingestion parser addresses #867 at the source: it checks compact-JSON key/value budgets before cloning, drops oversized entries whole, and stops after 32 accepted unknown fields. Known fields and rejected entries do not consume quota. The counting sink avoids an additional serialized buffer and rejects writes without budget underflow.

The current tests cover encoded-byte boundaries (including escapes and Unicode), the exact 139,329-byte saturated-map ceiling, ordinary extension round trips, persisted/reopened profiles, search-cache filtering, and fetch/refresh with a newer independently signed event. The latest commit also addresses the earlier test duplication and documentation concerns.

Verified current GitHub CI for this head: Required CI, all four Rust test shards, relay runtime tests, formatting, check, and clippy passed. Both PR commits have verified signatures and GitHub reports the PR clean and mergeable. Review used the remote diff and exact-head source; no local tests were run.

The documented exclusions are appropriate for this change: initial raw-event parsing remains unbounded by these retention limits, and historical oversized cache rows remain until newer metadata replaces them. These are not regressions introduced by this PR.
